packages/gatsby/src/utils/map-pages-to-static-query-hashes.ts

```diff
@@ -0,0 +1,174 @@
+ */
+export function mapTemplatesToStaticQueryHashes(
+  reduxState: IGatsbyState,
+  compilation
+): Map<string, Array<number>> {
+  /* The `staticQueryComponents` slice of state is useful because
+   * it is a pre extracted collection of all static queries found in a Gatsby site.
+   * This lets us traverse upwards from those to templates that
+   * may contain components that contain them.
+   * Note that this upward traversal is much shallower (and hence more performant)
+   * than an equivalent downward one from an entry point.
+   */
+  const { components, staticQueryComponents } = reduxState
+  const { modules } = compilation
+
+  /* When we traverse upwards, we need to know where to stop. We'll call these terminal nodes.
+   * `async-requires.js` is the entry point for every page, while `api-runner-browser-plugins.js`
+   * is the one for `gatsby-browser` (where one would use wrapRootElement or wrapPageElement APIs)
+   */
+  const terminalNodes = [
+    `.cache/api-runner-browser-plugins.js`,
+    `.cache/async-requires.js`,
+  ]
+
+  /* We call the queries included above a page (via wrapRootElement or wrapPageElement APIs)
+   * global queries. For now, we include these in every single page for simplicity. Overhead
+   * here is not much since we are storing hashes (that reference separate result files)
+   * as opposed to inlining results. We may move these to app-data perhaps in the future.
+   */
+  const globalStaticQueries = new Set<string>()
+
+  /* This function takes a webpack module corresponding
+   * to the file containing a static query and returns
+   * a Set of strings, each an absolute path of a dependent
+   * of this module
+   */
+  const getDeps = (mod): any => {
+    const staticQueryModuleComponentPath = mod.resource
+    const result = new Set()
+
+    // This is the body of the recursively called function
+    const getDepsFn = (m): any => {
+      // Reasons in webpack are literally reasons of why this module was included in the tree
+      const hasReasons = m.hasReasons()
+
+      // Is this node one of our known terminal nodes? See explanation above
+      const isTerminalNode = terminalNodes.reduce(
+        (result, terminalNode) => result || m?.resource?.includes(terminalNode),
+        false
+      )
+
+      // Exit if we don't have any reasons or we have reached a possible terminal node
+      if (!hasReasons || isTerminalNode) {
+        return result
+      }
+
+      // These are non terminal dependents and hence modules that need
+      // further upward traversal
+      const nonTerminalDependents = m.reasons
+        .filter(r => {
+          const dependentModule = r.module
+          const isTerminal = terminalNodes.reduce(
+            (result, terminalNode) =>
+              result || dependentModule?.resource?.includes(terminalNode),
+            false
+          )
+          return !isTerminal
+        })
+        .map(r => r.module)
+        .filter(Boolean)
```

Comment:
This is rather expensive.. You have 3 loops here. Could you reduce it to one loop?

Reply:
Good point. Unwound them!